stdlib/email/src/main/java/org/ballerinalang/stdlib/email/server/EmailConsumer.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.ballerinalang.stdlib.email.server;
+
+import org.ballerinalang.jvm.BallerinaErrors;
+import org.ballerinalang.jvm.BallerinaValues;
+import org.ballerinalang.jvm.values.ErrorValue;
+import org.ballerinalang.jvm.values.MapValue;
+import org.ballerinalang.jvm.values.ObjectValue;
+import org.ballerinalang.stdlib.email.client.EmailAccessClient;
+import org.ballerinalang.stdlib.email.util.EmailConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * Provides the capability to read an email and forward it to a listener.
+ *
+ * @since 1.3.0
+ */
+public class EmailConsumer {
+
+    private static final Logger log = LoggerFactory.getLogger(EmailConsumer.class);
+
+    private EmailListener emailListener;
+    private ObjectValue client;
+
+    /**
+     * Constructor for the EmailConsumer.
+     *
+     * @param emailProperties Map of property values
+     * @param listener Forwards the received emails to Ballerina code
+     */
+    public EmailConsumer(Map<String, Object> emailProperties, EmailListener listener) {
+        this.emailListener = listener;
+        String host = (String) emailProperties.get(EmailConstants.PROPS_HOST);
+        String username = (String) emailProperties.get(EmailConstants.PROPS_USERNAME);
+        String password = (String) emailProperties.get(EmailConstants.PROPS_PASSWORD);
+        String protocol = (String) emailProperties.get(EmailConstants.PROPS_PROTOCOL);
+        MapValue protocolConfig = (MapValue) emailProperties.get(EmailConstants.PROTOCOL_CONFIG);
+        if (protocol.equals(EmailConstants.IMAP)) {
+            client = BallerinaValues.createObjectValue(EmailConstants.EMAIL_PACKAGE_ID, EmailConstants.IMAP_CLIENT,
+                    host, username, password, (MapValue<Object, Object>) protocolConfig);
+            EmailAccessClient.initImapClientEndpoint(client, host, username, password,
+                    (MapValue<Object, Object>) protocolConfig);
+        } else if (protocol.equals(EmailConstants.POP)) {
+            client = BallerinaValues.createObjectValue(EmailConstants.EMAIL_PACKAGE_ID, EmailConstants.POP_CLIENT,
+                    host, username, password, (MapValue<Object, Object>) protocolConfig);
+            EmailAccessClient.initPopClientEndpoint(client, host, username, password,
+                    (MapValue<Object, Object>) protocolConfig);
+        } else {
+            String errorMsg = "Protocol should either be 'IMAP' or 'POP'.";
+            final EmailConnectorException e = new EmailConnectorException(errorMsg);
+            emailListener.onError(BallerinaErrors.createError(e));
```

Comment:
Since we initializing Email Consumer in the email listener lifecycle, we need to return an error and allow the system to panic. The error in configuration can't be recovered

Reply:
Done.